src/components/shared/MarkerTooltipContents.js

```diff
@@ -96,18 +96,90 @@ function _makePhaseTimesArray(
  * For example, it isn't useful to say "sweeping took 200ms" but it is
  * useful to say "sweeping.compacting took 20ms" (if compacting has no
  * sub-phases.
+ *
+ * We find leaf phases by constructing a tree of phase times and then
+ * reading its leaves.
  */
-function _isLeafPhase(
-  phaseName: string,
-  phases: Array<PhaseTimeTuple>
-): boolean {
-  return phases.every(
-    pt =>
-      // This phase is not the start of some other phase
-      !pt.name.startsWith(phaseName) ||
-      // Except itself.
-      pt.name === phaseName
-  );
+
+type PhaseTreeNode = PhaseTreeLeafNode | PhaseTreeBranchNode;
+
+type PhaseTreeLeafNode = {
+  node_type: 'leaf',
+  value: PhaseTimeTuple,
+};
+
+type PhaseTreeBranchNode = {
+  node_type: 'branch',
+  branches: Map<string, PhaseTreeNode>,
+};
+
+function _treeInsert(
+  tree: Map<string, PhaseTreeNode>,
+  path: Array<string>,
+  phase: PhaseTimeTuple
+) {
+  const component = path.shift();
+  if (component === undefined) {
+    // This path is not a leaf, it can be ignored.
+  } else {
+    const node = tree.get(component);
+    if (node) {
+      let branch_node;
+      switch (node.node_type) {
+        case 'leaf':
+          if (path.length > 0) {
+            // There are more path components, we need to convert this node
+            // into a branch and keep going.  We can delete the old leaf
+            // info.
+            const temp = (node: Object);
+            delete temp.value;
+            temp.node_type = 'branch';
+            temp.branches = new Map();
+            branch_node = (temp: PhaseTreeBranchNode);
+          } else {
+            // Duplicate leaves.
+            throw 'Duplicate phases';
+          }
+          break;
+        case 'branch':
+          branch_node = node;
+          break;
+        default:
+          throw 'Unreachable case'; // Tell flow that everything's okay.
```

Comment:
Hi @julienw 

I don't know what you mean by "(node.node_type: empty)". I tried replacing the default case with a case like:

        case empty:
          // Tell flow that everything's okay.
          throw new Error('Unreachable case');

But flow still seemed to think that control could fall through (because now there's no default case). I think I'm not clear on where to put (node.node_type: empty) or how it works.

Thanks.

Reply:
Doing:
        default:
          (node.node_type : empty);

seemed to make no difference.  What I'd like is for flow to understand that there ere exactly two options here. the default case will never be executed.